akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -148,18 +179,21 @@ import scala.util.control.NonFatal
     } else rec(s)
   }
 
-  private def establishSubscription(subscriber: Subscriber[_], subscription: Subscription): Unit = {
+  private def establishSubscription(establishing: Establishing, subscription: Subscription): Unit = {
     val wrapped = new WrappedSubscription(subscription)
     try {
-      subscriber.onSubscribe(wrapped)
+      if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode.establishSubscription(wrapped)")
+      establishing.sub.onSubscribe(wrapped)
       // Requests will be only allowed once onSubscribe has returned to avoid reentering on an onNext before
       // onSubscribe completed
       wrapped.ungateDemandAndRequestBuffered()
+      if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode.establishSubscription(wrapped) -> Both")
+      set(Both(establishing.sub)) // only place we transition from establishing
```

Comment:
it can also be (already) in Both, but I guess that is fine

Reply:
unless it could also have moved to Inert and we are setting it back to Both by this?

Reply:
It should only be able to go to `Both` by first being `Establishing`, if it isn't I made a mistake somewhere

Reply:
if you follow who is calling `establishSubscription` it's one case where it's already `Both` and before that `Subscriber`

Reply:
I found that one now while rewriting to not do the spin. So fixed in a push soon.